File: 프로그래머스/lv2/17687. ［3차］ n진수 게임/［3차］ n진수 게임.java
import java.util.*;

class Solution {

    StringBuilder sb;
    ArrayList<Character> answers;

    public String solution(int n, int t, int m, int p) {
        sb = new StringBuilder();
        answers = new ArrayList<>();
        Integer number = 0;

        while (answers.size() <= t * m) {
            String tmp = changeNumber(number, n);
            for (char ch : tmp.toCharArray()) {
                answers.add(ch);
            }
            number += 1;
        }

        int idx = 0;
        int turn = 1;
        while (sb.length() < t) {
            if (turn == p) {
                char ch = answers.get(idx);
                if (ch >= 'a' && ch <= 'z') {
                    sb.append(Character.toUpperCase(ch));
                } else
                    sb.append(answers.get(idx));
            }

            if (turn == m) {
                turn = 0;
            }

            turn += 1;
            idx += 1;
        }

        return sb.toString();
    }

    public String changeNumber(Integer number, int n) {
        return Integer.toString(number, n);
    }
}

File: 프로그래머스/lv2/17687. ［3차］ n진수 게임/［3차］ n진수 게임.js
const buildOrderString = (n, t, m) => {
    let answer = '', number = 0;

    while (answer.length <= t * m) {
        let buffer = number.toString(n);

        for (let ch of buffer) answer += ch;
        
        number++;
    }

    return answer;
}

const buildAnswerString = (order, t, m, p) => {
    let answer = '', idx = p;

    while (idx < order.length && answer.length < t) {
        let ch = order[idx - 1];

        if (ch >= 'a' && ch <= 'z') {
            ch = String(ch).toUpperCase();
        }

        answer += ch;
        idx += m;
    }

    return answer;
}



const solution = (n, t, m, p) => {
    let order = buildOrderString(n, t, m);
    return buildAnswerString(order, t, m, p);
}


